[lld]Add lld/Common/BPSectionOrdererBase from MachO for reuse in ELF

Colibrow opened the pull request:
Add lld/Common/BPSectionOrdererBase from MachO for reuse in ELF

ellishg commented:
Thanks for the PR! I'll try to take a look, but I might not get to it until next week because Thursday is Thanksgiving. I'll also assign a few reviewers that might be interested in taking a look.

smithp35 commented:
Just a quick drive by comment on the ELF side, I've not had a chance to go through the code in detail. I think the .subsections_via_symbols is MachO specific. Not quite sure what it does in ELF. I think the equivalent in clang at least is -ffunction-sections and for data -fdata-sections. For assembler the user has to be careful to make a section per function.

If I'm right it could be worth updating comments and the assembler test.

Colibrow commented:
@smithp35 You're correct. In the standard ELF format, we typically use -ffunction-sections and -fdata-sections flags to achieve balanced section partitioning. The original comment was indeed copied from the Mach-O specific implementation. I've updated the code to reflect this, and if you have any additional suggestions, I'd be happy to hear them. Appreciate your review!

mysterymath commented on lld/ELF/Options.td:
```diff
@@ -141,6 +141,20 @@ def call_graph_profile_sort: JJ<"call-graph-profile-sort=">,
 def : FF<"no-call-graph-profile-sort">, Alias<call_graph_profile_sort>, AliasArgs<["none"]>,
   Flags<[HelpHidden]>;
 
+defm irpgo_profile_sort: Eq<"irpgo-profile-sort",
```
On the embedded side, we've discussed consuming profiles for an alternative goal: automatically placing hotter sections in better regions of memory.

Would it be possible to separate out the flag that provides the profile, say, `--irpgo_profile`, from a flag that enables sorting for startup time, say, `--startup-sort`? There's already a small family of sort concerns present in this patch, and that organization would support adding the embedded one in-kind once we iron out the details.

Colibrow replied:
LGTM. 
How about this separation in Mach-O, as it is currently used in both ELF and Mach-O? Or implement it just in ELF in this commit?

MaskRay replied:
Ideally implement the change for Mach-O in a prerequisite patch, then implement it in this PR as a follow-up

guideline: improve the base, then port the base

Colibrow replied:
@MaskRay @ellishg @mysterymath I have created a new pull request about the options change. Could you please review it?

https://github.com/llvm/llvm-project/pull/118594
If it works I'll rebase this one.

MaskRay commented on lld/ELF/Driver.cpp:
```diff
@@ -1255,6 +1255,45 @@ static void readConfigs(Ctx &ctx, opt::InputArgList &args) {
+      ctx.arg.functionOrderForCompression = true;
+      ctx.arg.dataOrderForCompression = true;
+    } else if (compressionSortStr != "none") {
+      ErrAlways(ctx) << "unknown value `" + compressionSortStr + "` for " +
```
we use single quotes

MaskRay commented on lld/ELF/BPSectionOrderer.cpp:
```diff
@@ -0,0 +1,71 @@
+  size_t highestAvailablePriority = std::numeric_limits<int>::max();
+  SmallVector<lld::BPSectionBase *> sections;
+
+  for (Symbol *sym : ctx.symtab->getSymbols()) {
```
all braces in this nest are unneeded https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

MaskRay replied:
Needs a comment about what `sections` is

MaskRay commented on lld/ELF/BPSectionOrderer.cpp:
```diff
@@ -0,0 +1,71 @@
+    }
+  }
+
+  for (ELFFileBase *file : ctx.objectFiles)
```
ditto

MaskRay commented on lld/ELF/BPSectionOrderer.h:
```diff
@@ -0,0 +1,140 @@
+
+class InputSection;
+
+class ELFSymbol : public BPSymbol {
```
This name is generic and can lead to confusion. ISTM `BP` should be part of the name, perhaps BPSymbolELF
(ala MCSectionELF)

MaskRay commented on lld/ELF/BPSectionOrderer.h:
```diff
@@ -0,0 +1,140 @@
+class ELFSection : public BPSectionBase {
+  const InputSectionBase *isec;
+  ELFSymbol *symbol;
+  std::vector<BPSymbol *> symbols;
```
`SmallVector<, 0>` to make the struct smaller

MaskRay commented on lld/ELF/BPSectionOrderer.h:
```diff
@@ -0,0 +1,140 @@
+    }
+
+    // Calculate content hashes
+    for (size_t i = 0; i < isec->content().size(); i++) {
```
https://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop

I think many places in this patch violate this and should be fixed

MaskRay commented on lld/ELF/CMakeLists.txt:
```diff
@@ -56,6 +56,7 @@ add_lld_library(lldELF
   SymbolTable.cpp
   Symbols.cpp
   SyntheticSections.cpp
+  BPSectionOrderer.cpp
```
alphabetical order

MaskRay commented on lld/ELF/Options.td:
```diff
@@ -141,6 +141,20 @@ def call_graph_profile_sort: JJ<"call-graph-profile-sort=">,
+    "Order startup functions to improve compressed size in addition to startup time",
+    "Do not order startup function for compression">;
+
+def compression_sort: JJ<"compression-sort=">,
```
@ellishg ELF has the compressed section feature https://maskray.me/blog/2022-01-23-compressed-debug-sections

Mach-O doesn't have the feature, but would it make sense to rename the Mach-O option? Perhaps these BP related options should all use bp- as a prefix, unless obvious from the context (e.g. `--irpgo-*`)?

ellishg replied:
Yeah we can rename these options to `-bp-compression-sort-startup-functions` and `-bp-compression-sort`.

MaskRay commented on lld/test/ELF/bp-section-orderer-errs.s:
```diff
@@ -0,0 +1,44 @@
+# REQUIRES: aarch64
+
+# RUN: llvm-mc -filetype=obj -triple=aarch64 %s -o %t
```
`rm -rf %t && split-file %s %t && cd %t`

delete `-o /dev/null` for not ld.lld tests

See `2d756d9d4c89ac50404f942f2db2f4a402aa0e00`

MaskRay commented on lld/test/ELF/bp-section-orderer.s:
```diff
@@ -0,0 +1,336 @@
+# REQUIRES: aarch64
+
+# RUN: rm -rf %t && split-file %s %t
```
append `&& cd %t` to avoid `%t` complexity below.

MaskRay commented on lld/test/ELF/bp-section-orderer.s:
```diff
@@ -0,0 +1,336 @@
+# RUN: llvm-mc -filetype=obj -triple=aarch64 %t/a.s -o %t/a.o
+# RUN: llvm-profdata merge %t/a.proftext -o %t/a.profdata
+
+# The code below is the reduced version of the output
```
See https://llvm.org/docs/TestingGuide.html#elaborated-tests

you may want `-fno-exceptions -fno-rtti -fno-asynchronous-unwind-tables`

MaskRay reviewed (changes requested):
.

ellishg commented on lld/MachO/BPSectionOrderer.h:
```diff
@@ -14,13 +14,149 @@
+    if (auto *sym = reloc.referent.dyn_cast<Symbol *>()) {
+      kind += (" Symbol " + Twine(sym->kind())).str();
+      if (auto *d = llvm::dyn_cast<Defined>(sym)) {
+        if (llvm::isa_and_nonnull<CStringInputSection>(isec))
+          return BPSectionBase::getRelocHash(kind, 0, isec->getOffset(d->value),
+                                             reloc.addend);
```
We no longer check if this section is a cstring section. https://github.com/llvm/llvm-project/commit/6827a00d4dfe1fa7d479337b6192602744055686

ellishg commented on lld/test/ELF/bp-section-orderer-errs.s:
```diff
@@ -0,0 +1,45 @@
+# REQUIRES: aarch64
```
I think you can move all these "is incompatible with" tests to https://github.com/llvm/llvm-project/blob/main/lld/test/ELF/incompatible.s
Just add `touch %t.call_graph` and `touch %t.profdata` to create the files. They won't be parsed so they can be empty. You also won't need `split-file`. That will remove a lot of the complexity of this test.

ellishg commented on lld/test/ELF/bp-section-orderer-stress.s:
```diff
@@ -0,0 +1,104 @@
+# REQUIRES: aarch64
```
I wish we didn't have to duplicate this test in ELF and Mach-O, but I can't think of a better solution right now.

MaskRay replied:
Duplication like this fine as sharing the test could make updating either port more difficult.

ellishg commented on lld/test/ELF/bp-section-orderer.s:
```diff
@@ -0,0 +1,868 @@
+}
+#--- gen
+echo '#--- a.s'
+clang -target aarch64-linux-gnu -fdebug-compilation-dir='/proc/self/cwd' -ffunction-sections -fdata-sections -fno-exceptions -fno-rtti -fno-asynchronous-unwind-tables -S -g a.cc -o -
```
I don't believe debug info is required for this test and it will greatly reduce the size of the generated assembly.

```suggestion
clang -target aarch64-linux-gnu -fdebug-compilation-dir='/proc/self/cwd' -ffunction-sections -fdata-sections -fno-exceptions -fno-rtti -fno-asynchronous-unwind-tables -S a.cc -o -
```

ellishg commented on lld/test/ELF/bp-section-orderer.s:
```diff
@@ -0,0 +1,868 @@
+	.section	.debug_line,"",@progbits
+.Lline_table_start0:
+
+# RUN: ld.lld -e main -o a.out a.o --irpgo-profile-sort=a.profdata --verbose-bp-section-orderer 2>&1 | FileCheck %s --check-prefix=STARTUP
+# RUN: ld.lld -e main -o a.out a.o --irpgo-profile-sort=a.profdata --verbose-bp-section-orderer --icf=all --compression-sort=none 2>&1 | FileCheck %s --check-prefix=STARTUP
+
+# STARTUP: Ordered 3 sections using balanced partitioning
+
+# RUN: ld.lld -e main -o - a.o --irpgo-profile-sort=a.profdata --symbol-ordering-file a.orderfile | llvm-nm --numeric-sort --format=just-symbols - | FileCheck %s --check-prefix=ORDERFILE
```
I think these lines and all the following lines should go at the top of the file. Anything after the `clang` command should be generated by that command.

> The script will prepare extra files with split-file, invoke gen, and then rewrite the part after gen with its stdout.

https://llvm.org/docs/TestingGuide.html#elaborated-tests

ellishg commented on lld/test/ELF/bp-section-orderer.s:
```diff
@@ -0,0 +1,868 @@
+# RUN: llvm-mc -filetype=obj -triple=aarch64 a.s -o a.o
+# RUN: llvm-profdata merge a.proftext -o a.profdata
+
+.ifdef GEN
```
I think you cannot use the simplified `.ifdef GEN` style if you are running more commands in this test.

thevinster commented on lld/Common/BPSectionOrdererBase.cpp:
```diff
@@ -0,0 +1,383 @@
+    bool forFunctionCompression, bool forDataCompression,
+    bool compressionSortStartupFunctions, bool verbose,
+    SmallVector<BPSectionBase *> inputSections) {
+  TimeTraceScope timeScope("Balanced Partitioning");
```
Might want to rename this to something like "Setup Balanaced Partitioning" since there is already a timeScope for "Balanced Partitioning" below.

thevinster reviewed (commented):
Some nits on top of everything else.

Colibrow commented on lld/MachO/BPSectionOrderer.cpp:
```diff
@@ -11,425 +11,41 @@
-  SmallVector<const InputSection *> sections;
-  DenseMap<const InputSection *, uint64_t> sectionToIdx;
-  StringMap<DenseSet<unsigned>> symbolToSectionIdxs;
+  SmallVector<BPSectionBase *> sections;
```
I'm not sure how to pass the sections into the `reorderSectionsByBalancedPartitioning` function which locates in lld/common since the BPSectionELF and BPSectionMacho both need to be accepted. now I use the `SmallVector<std::unique_ptr<BPSectionBase>>` and `SmallVector<std::unique_ptr<BPSectionMacho>>`. Could you give some more details?

Colibrow commented on lld/Common/BPSectionOrdererBase.cpp:
```diff
@@ -0,0 +1,383 @@
+    // Mangler::getNameWithPrefix() for details.
+    if (name.consume_front("_") || name.consume_front("l_"))
+      rootSymbolToSectionIdxs[name].insert(sectionIdxs.begin(),
+                                           sectionIdxs.end());
```
I created a new virtual function in BPSectionBase and implemented it for both ELF and Mach-O sections. I'm calling it like` isec->needResolveLinkageName(rootSymbol)`. Are there more elegant ways to approach this design?

Colibrow commented:
@ellishg @MaskRay Hi, I've rebased this PR with commit [118594](https://github.com/llvm/llvm-project/pull/118594) and addressed the code review comments. Could you please take another look?

ellishg commented on lld/MachO/BPSectionOrderer.h:
```diff
@@ -14,13 +14,140 @@
+
+  // Linkage names can be prefixed with "_" or "l_" on Mach-O. See
+  // Mangler::getNameWithPrefix() for details.
+  bool needResolveLinkageName(llvm::StringRef &name) const override {
+    return (name.consume_front("_") || name.consume_front("l_"));
+  }
```
I see now that `StringRef` is passed by reference, but this is confusing and easy to miss. Instead, what do you think about returning an optional stringref?

```suggestion
  std::optional<StringRef> getResolvedLinkageName(llvm::StringRef name) const override {
    if (name.consume_front("_") || name.consume_front("l_"))
      return name;
    return {};
  }
```

ellishg commented on lld/ELF/BPSectionOrderer.h:
```diff
@@ -0,0 +1,140 @@
+
+  void getSectionHash(llvm::SmallVectorImpl<uint64_t> &hashes,
+                      const llvm::DenseMap<const BPSectionBase *, uint64_t>
+                          &sectionToIdx) const override {
```
@Colibrow What do you think of this?

ellishg reviewed (changes requested):
Don't forget to change the flag spelling in the PR summary

Colibrow commented:
@ellishg I deleted the function that passed the map everywhere, and the file [lld/ELF/BPSectionOrderer.h](https://github.com/llvm/llvm-project/pull/117514/files#diff-f2b97e0dddb816b7e1df8c063a72dd4c87ea80ee76795b5d44bc6a32e4538330) has been updated. Does this refer to the previous implementation?

ellishg commented on lld/ELF/Writer.cpp:
```diff
@@ -1083,6 +1084,15 @@ static void maybeShuffle(Ctx &ctx,
+      ctx.arg.bpDataOrderForCompression) {
+    TimeTraceScope timeScope("Balanced Partitioning Section Orderer");
+    sectionOrder = runBalancedPartitioning(
+        ctx, ctx.arg.irpgoProfilePath, ctx.arg.bpFunctionOrderForCompression,
```
```suggestion
        ctx, ctx.arg.bpStartupFunctionSort ? ctx.arg.irpgoProfilePath : "", ctx.arg.bpFunctionOrderForCompression,
```

ellishg commented on lld/ELF/BPSectionOrderer.h:
```diff
@@ -0,0 +1,133 @@
+  }
+
+  InputSectionBase *getInputSection() const {
+    if (auto *d = llvm::dyn_cast<Defined>(sym))
```
```suggestion
    if (auto *d = asDefined())
```

ellishg commented on lld/ELF/BPSectionOrderer.h:
```diff
@@ -0,0 +1,133 @@
+
+  SmallVector<std::unique_ptr<BPSymbol>> getSymbols() const override {
+    SmallVector<std::unique_ptr<BPSymbol>> symbols;
+    if (auto *d = llvm::dyn_cast_or_null<Defined>(symbol))
```
If you make `symbol` type `BPSymbolELF *` you can do this
```suggestion
    if (auto *d = symbol->asDefined())
```

ellishg reviewed (commented):
It's looking better! After we resolve the section index issue in `getRelocHash()` I want to test this on my codebase to make sure we don't have a regression.

Colibrow commented on lld/MachO/BPSectionOrderer.h:
```diff
@@ -14,13 +14,138 @@
+    auto *isec = reloc.getReferentInputSection();
+    std::optional<uint64_t> sectionIdx;
+    if (isec && isec == section->getSection())
+      sectionIdx = section->getSectionIdx();
```
Yes. Added the sectiontToIdx map.

ellishg commented on lld/MachO/BPSectionOrderer.h:
```diff
@@ -123,13 +124,17 @@ class BPSectionMacho : public BPSectionBase {
-    if (isec && isec == section->getSection())
-      sectionIdx = section->getSectionIdx();
-
+    for (const auto &entry : sectionToIdx)
+      if (const auto *bpSection = llvm::dyn_cast<BPSectionMacho>(entry.first))
+        if (bpSection->getSection() == isec) {
+          sectionIdx = entry.second;
+          break;
+        }
```
This seems to do what we want, but it is very inefficient because we are iterating over the whole map. We should instead use `DenseMap<const InputSection *, uint64_t>` so we can simply use `sectionToIdx.find(isec)`.

Colibrow replied:
I used sectionToIdx.find(isec) to handle this, but since sectionToIdx is a DenseMap<const BPSectionBase *, uint64_t> in BPSectionOrdererBase.cpp, I updated it accordingly. Is there a better way to write this?

ellishg replied:
Yeah the key type needs to be 'InputSection'

ellishg commented on lld/MachO/BPSectionOrderer.h:
```diff
@@ -14,13 +14,145 @@
+                            &sectionToIdx) const override {
+    constexpr unsigned windowSize = 4;
+
+    // Create a map from BPSectionBase* to InputSection* for lookup
+    llvm::DenseMap<const InputSection *, uint64_t> isecToIdx;
+    for (const auto &[section, idx] : sectionToIdx)
+      if (auto *machoSection = llvm::dyn_cast<BPSectionMacho>(section))
+        isecToIdx[machoSection->getSection()] = idx;
```
Sorry, this is still not efficient because we call `getSectionHashes` for each section, making this `O(n^2)`.

I guess we only need `sectionToIdx` for Mach-O because ELF doesn't take relocations into account. Can we somehow make `DenseMap<const InputSection *, uint64_t> sectionToIdx` specific to Mach-O? Or should we make the key type `void *` and cast to the Mach-O or ELF version of `InputSection *` when we need it?

By the way, have you tried to hash relocations on ELF? My guess is that you could get a pretty nice uncompressed and compressed size win.

Colibrow replied:
I'm trying to use void * for this. Based on our MachO hash relocation results, I think it will perform quite well on ELF too. I'll spend some time refining it.

Colibrow replied:
@ellishg Hi ellis, considering the complexity of implementing ELF relocations, I think the hash relocations for ELF may take some time to complete. How about using the simple sectionHash as a base for now and adding relocation optimizations in the future? What do you think?

ellishg replied:
I'm fine with that for ELF, but we don't want to regress anything for Mach-O

Colibrow replied:
Got it. I'll ensure there are no regressions for Mach-O while doing ELF relocations hashes.

Colibrow commented on lld/ELF/BPSectionOrderer.h:
```diff
@@ -102,16 +102,41 @@ class BPSectionELF : public BPSectionBase {
-    // TODO: Calculate relocation hashes.
-    // Since in ELF, relocations are complex, but the effect without them are
-    // good enough, we just use 0 as their hash.
+    for (const auto &r : isec->relocations) {
```
@ellishg @MaskRay Hi, I've updated the relocations hashes by the `isec->relocations` which seems an common interface to access relocation. Could you please review these?

ellishg replied:
Nice, it seems good. Did you see an uncompressed or compressed size win with this change?

ellishg commented:
I built this PR locally and I needed these cmake changes for it to work and some of the includes are not necessary

```diff
diff --git a/lld/Common/BPSectionOrdererBase.cpp b/lld/Common/BPSectionOrdererBase.cpp
index f7b460d51a16..51d56606ec92 100644
--- a/lld/Common/BPSectionOrdererBase.cpp
+++ b/lld/Common/BPSectionOrdererBase.cpp
@@ -19,7 +19,6 @@
 #include "llvm/Support/BalancedPartitioning.h"
 #include "llvm/Support/TimeProfiler.h"
 #include "llvm/Support/VirtualFileSystem.h"
-#include "llvm/Support/xxhash.h"

 #define DEBUG_TYPE "bp-section-orderer"
 using namespace llvm;
diff --git a/lld/Common/CMakeLists.txt b/lld/Common/CMakeLists.txt
index 2ab5093bf688..43e91b85821d 100644
--- a/lld/Common/CMakeLists.txt
+++ b/lld/Common/CMakeLists.txt
@@ -48,6 +48,7 @@ add_lld_library(lldCommon
   Demangle
   MC
   Option
+  ProfileData
   Support
   Target
   TargetParser
diff --git a/lld/ELF/BPSectionOrderer.cpp b/lld/ELF/BPSectionOrderer.cpp
index 86605b1a7626..1fc6036e5dd9 100644
--- a/lld/ELF/BPSectionOrderer.cpp
+++ b/lld/ELF/BPSectionOrderer.cpp
@@ -11,10 +11,7 @@
 #include "InputFiles.h"
 #include "InputSection.h"
 #include "lld/Common/BPSectionOrdererBase.h"
-#include "lld/Common/CommonLinkerContext.h"
 #include "llvm/ADT/DenseMap.h"
-#include "llvm/Support/BalancedPartitioning.h"
-#include "llvm/Support/TimeProfiler.h"

 #include "SymbolTable.h"
 #include "Symbols.h"
diff --git a/lld/ELF/BPSectionOrderer.h b/lld/ELF/BPSectionOrderer.h
index 7f32c26bdd3e..502b7cfbdd2f 100644
--- a/lld/ELF/BPSectionOrderer.h
+++ b/lld/ELF/BPSectionOrderer.h
@@ -21,7 +21,6 @@
 #include "lld/Common/BPSectionOrdererBase.h"
 #include "llvm/ADT/DenseMap.h"
 #include "llvm/ADT/STLExtras.h"
-#include "llvm/ADT/STLFunctionalExtras.h"
 #include "llvm/ADT/StringRef.h"
 #include "llvm/BinaryFormat/ELF.h"
 #include "llvm/Support/ErrorHandling.h"
diff --git a/lld/ELF/CMakeLists.txt b/lld/ELF/CMakeLists.txt
index 298443cd6ea4..ec3f6382282b 100644
--- a/lld/ELF/CMakeLists.txt
+++ b/lld/ELF/CMakeLists.txt
@@ -73,6 +73,7 @@ add_lld_library(lldELF
   Object
   Option
   Passes
+  ProfileData
   Support
   TargetParser
   TransformUtils
diff --git a/lld/MachO/BPSectionOrderer.cpp b/lld/MachO/BPSectionOrderer.cpp
index 5e9fd5248b2e..0ffbf16007fd 100644
--- a/lld/MachO/BPSectionOrderer.cpp
+++ b/lld/MachO/BPSectionOrderer.cpp
@@ -8,11 +8,7 @@

 #include "BPSectionOrderer.h"
 #include "InputSection.h"
-#include "lld/Common/ErrorHandler.h"
 #include "llvm/ADT/DenseMap.h"
-#include "llvm/ADT/StringMap.h"
-#include "llvm/Support/BalancedPartitioning.h"
-#include "llvm/Support/TimeProfiler.h"

 #define DEBUG_TYPE "bp-section-orderer"

diff --git a/lld/MachO/BPSectionOrderer.h b/lld/MachO/BPSectionOrderer.h
index fa71f739ecb3..29b20c781c6b 100644
--- a/lld/MachO/BPSectionOrderer.h
+++ b/lld/MachO/BPSectionOrderer.h
@@ -20,7 +20,6 @@
 #include "lld/Common/BPSectionOrdererBase.h"
 #include "llvm/ADT/DenseMap.h"
 #include "llvm/ADT/StringRef.h"
-#include "llvm/ADT/TinyPtrVector.h"

 namespace lld::macho {

diff --git a/lld/include/lld/Common/BPSectionOrdererBase.h b/lld/include/lld/Common/BPSectionOrdererBase.h
index 78c51cd9786b..a22bfcafd79c 100644
--- a/lld/include/lld/Common/BPSectionOrdererBase.h
+++ b/lld/include/lld/Common/BPSectionOrdererBase.h
@@ -14,7 +14,6 @@
 #ifndef LLD_COMMON_BP_SECTION_ORDERER_BASE_H
 #define LLD_COMMON_BP_SECTION_ORDERER_BASE_H

-#include "llvm/ADT/ArrayRef.h"
 #include "llvm/ADT/DenseMap.h"
 #include "llvm/ADT/StringRef.h"
 #include "llvm/ADT/Twine.h"
```

ellishg commented on lld/ELF/CMakeLists.txt:
```diff
@@ -72,6 +73,7 @@ add_lld_library(lldELF
   Object
   Option
   Passes
+  ProfileData
```
Now that I think about it more, I think we only need this dependency because we use `InstrProfReader::`. But this is only used in `lld/Common`, so we might not need this explicit dependency here. Same for `MachO/CMakeLists.txt`, I think `ProfileData` can be removed there.

ellishg reviewed (approved):
LGTM! Thanks for ensuring this works for both targets! I tested that `MachO/bp-section-orderer-stress.s` produced the same order before and after that change, so it gives me some confidence that this won't regress anything. And I'm curious what size results you get for ELF, so please publish that when it is available.

MaskRay commented on lld/Common/BPSectionOrdererBase.cpp:
```diff
@@ -0,0 +1,378 @@
+using namespace llvm;
+using UtilityNodes = SmallVector<BPFunctionNode::UtilityNodeT>;
+
+namespace lld {
```
We don't use `name xxx { }` in .cpp files https://llvm.org/docs/CodingStandards.html#use-namespace-qualifiers-to-implement-previously-declared-functions

MaskRay commented on lld/Common/BPSectionOrdererBase.cpp:
```diff
@@ -0,0 +1,378 @@
+#include "llvm/Support/TimeProfiler.h"
+#include "llvm/Support/VirtualFileSystem.h"
+
+#define DEBUG_TYPE "bp-section-orderer"
```
blank line separator between #define and using

MaskRay commented on lld/include/lld/Common/BPSectionOrdererBase.h:
```diff
@@ -0,0 +1,77 @@
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/ADT/Twine.h"
+#include "llvm/Support/xxhash.h"
```
include <optional>

MaskRay reviewed (changes requested):
I need time to look at the ELF port changes.

If you drop ELF changes from this PR, and just move MachO code to Common to be reused, this PR will be good to go and I'll approve.

You need a dependent ELF patch.

Colibrow committed:
feat: move the BPSectionOrder from MachO to Common for reuse in ELF later

Colibrow commented:
@MaskRay I dropped the ELF parts and will post the ELF patch as a new PR once this PR is merged.

Colibrow commented:
@MaskRay Hi, Can you merge this PR? I don't have merge button for this repository.

lerno commented:
Could this have been causing this issue #122655? I've been hoping this would be discovered and fixed, but it's been broken now for almost a month.

MaskRay commented:
Spent more time on BP. I believe that the virtual classes as done by this PR `BPSectionBase`/`BPSymbol` is quite wasteful.   `BPSectionBase`/`BPSymbol`  are thin wrappers that contain just one pointer. This is quite wasteful.

Perhaps we should treat Symbol and InputSection as `void *` in BPSectionOrderer.cpp and use the following class hierarchy instead.

```
class BPBase {
  virtual getSectionSize(void *);
  virtual isCodeSection(void *);
  
  virtual StringRef getSymbolName(void *);
  virtual uint64_t getSymbolValue(void *);
};
class BPMachO;
class BPELF; // future
```

Then we can avoid allocating a large section and `inputSections` (expensive for large ELF applications).

We could also leverage Curiously Recurring Template Pattern if we want to retain type information of symbols and sections.

MaskRay commented:
@Colibrow I am taking a stab on the CRTP refactoring.

Colibrow commented:
I've been using the template method to implement the project, but I'm stuck. In the first version, I placed the base template in `lld/common`, and during compilation, the `lld/ELF` and `lld/MachO` templates couldn't find the symbols. I might have made a mistake in my implementation, and I'm hoping to learn how to refactor this correctly. Nice work.
